hoomd/write/hdf5.py

```diff
@@ -0,0 +1,272 @@
+        raise ValueError("Attribute {} is read-only.".format(attr))
+
+    def attach(self, simulation):
+        self._initialize(simulation.device.communicator)
+        self._attached_ = True
+
+    def _attached(self):
+        return self._attached_
+
+    def detach(self):
+        self._attached_ = False
+        if self._fh is not None:
+            self._fh.close()
+
+    def act(self, timestep):
+        """Write a new frame of logger data to the HDF5 file."""
+        log_dict = util._dict_flatten(self.logger.log())
+        if self._fh is None:
+            return
+        if self._frame == 0:
+            self._initialize_datasets(log_dict)
+        for key, (value, category) in log_dict.items():
+            if logging.LoggerCategories[category] in self._reject_categories:
+                continue
+            if value is None:
+                continue
+            str_key = "/".join(("hoomd-data",) + key)
+            if str_key not in self._fh:
+                raise RuntimeError(
+                    "The logged quantities cannot change within a file.")
+            dataset = self._fh[str_key]
+            dataset.resize(self._frame + 1, axis=0)
+            dataset[self._frame, ...] = value
+        self._frame += 1
+        self._fh["hoomd-data"].attrs["frames"] = self._frame
+
+    @_skip_fh
+    def flush(self):
+        """Write out all data currently buffered in memory.
+
+        Without calling this, data may be stored in the h5py.File object without
+        being written to the on disk file yet.
+        """
+        self._fh.flush()
+
+    @_skip_fh
+    def _create_dataset(self, key: str, shape, dtype, chunk_size):
+        self._fh.create_dataset(
+            key,
+            shape,
+            dtype=dtype,
+            chunks=chunk_size,
+            maxshape=(None,) + shape[1:],
+        )
+
+    @_skip_fh
+    def _initialize_datasets(self, log_dict):
+        """Create datasets setting shape, dtype, and chunk size.
+
+        Chunk size does not seem to matter much to the write performance of the
+        file surprisingly. However, it has a drastic effect on reads. The
+        current settings balance file size (smaller scalar chunks) with reading
+        speed.
+
+        Tests were done on 1,000 frame files for writes and reads and tested for
+        writing and reading speed.
+        """
+        for key, (value, category) in log_dict.items():
+            chunk_size = None
+            if category == "scalar":
+                data_shape = (1,)
+                dtype = "f8" if isinstance(value, float) else "i8"
+                chunk_size = (500,)
+            else:
+                if not isinstance(value, np.ndarray):
+                    value = np.asarray(value)
+                data_shape = (1,) + value.shape
+                dtype = value.dtype
+                chunk_size = [dim if dim <= 10 else dim for dim in data_shape]
+                chunk_size[0] = 1
```

Comment:
```suggestion
                chunk_size = data_shape
```
Unless this code is an error and you intend to perform some modification to the chunk size in the list comprehension.

Also, I would assume that `chunk_size = data_shape` works very well for typical per-particle, per-bond, and other large arrays. It might not work so well for fixed length 2 arrays. For parity with the scalar chunk size above, try:
```
chunk_size[0] = max(4096 // value.nbytes, 1)
```
If this works well, remove magic numbers by making 4096 a named constant and using it in both places we compute a chunk size in dimensions.

Reply:
I originally did, but it did not change performance.